File: frontend/src/api/workflow.ts
import { Node, TopologyTemplate, Connection, IfElseCondition } from '../types';

const API_BASE = '/api';

interface WorkflowData {
  id?: string;
  name: string;
  agents: AgentData[];
  topologies: TopologyData[];
  connections: ConnectionData[];
  nodes: WorkflowNodeData[];
}

interface AgentData {
  id: string;
  name: string;
  instructions: string;
  model: string;
  topology_id: string | null;
  topology_role: string | null;
}

interface TopologyData {
  id: string;
  type: string;
  name: string;
  agents: string[];
  internal_edges: InternalEdgeData[];
  max_turns: number;
  timeout: number;
  early_termination: boolean;
  start_agent_id: string | null;
}

interface InternalEdgeData {
  id: string;
  from: string;
  to: string;
  type: string;
}

interface ConnectionData {
  id: string;
  from_id: string;
  to_id: string;
  from_type: string;
  from_port?: number;
}

interface WorkflowNodeData {
  id: string;
  type: 'start' | 'end' | 'agent' | 'topology' | 'ifelse' | 'approval' | 'note';
  name: string;
  config: Record<string, unknown>;
  conditions?: Array<{ id: string; name?: string; condition: string }>;
  approval_message?: string;
}

interface ApprovalRequiredData {
  execution_id: string;
  node_id: string;
  message: string;
}

interface ExecutionMessage {
  id: string;
  from: string;
  to: string;
  content: string;
  timestamp: string;
  metadata: Record<string, unknown>;
}

interface ConversationMessage {
  role: 'user' | 'assistant';
  content: string;
}

// Convert frontend state to API format
export const convertToApiFormat = (
  nodes: Node[],
  topologyTemplates: TopologyTemplate[],
  connections: Connection[],
  workflowName: string = 'Untitled Workflow'
): WorkflowData => {
  // Convert agents
  const agents: AgentData[] = nodes
    .filter(n => n.type === 'agent')
    .map(n => ({
      id: n.id,
      name: n.name,
      instructions: (n.config?.instructions as string) || '',
      model: (n.config?.model as string) || 'gpt-4o',
      topology_id: n.topologyId,
      topology_role: n.topologyRole,
    }));

  // Convert topologies
  const topologies: TopologyData[] = topologyTemplates.map(t => ({
    id: t.id,
    type: t.type,
    name: t.name,
    agents: nodes.filter(n => n.topologyId === t.id).map(n => n.id),
    internal_edges: t.internalEdges.map(e => ({
      id: e.id,
      from: e.from,
      to: e.to,
      type: e.type,
    })),
    max_turns: t.maxTurns,
    timeout: t.timeout,
    early_termination: t.earlyTermination,
    start_agent_id: t.startAgentId,
  }));

  // Convert connections (including from_port for branching)
  const connectionData: ConnectionData[] = connections.map(c => ({
    id: c.id,
    from_id: c.from,
    to_id: c.to,
    from_type: c.fromType,
    from_port: c.fromPort,
  }));

  // Convert workflow nodes (start, end, ifelse, approval, note)
  const workflowNodes: WorkflowNodeData[] = nodes
    .filter(n => ['start', 'end', 'ifelse', 'approval', 'note'].includes(n.type))
    .map(n => {
      const nodeData: WorkflowNodeData = {
        id: n.id,
        type: n.type as WorkflowNodeData['type'],
        name: n.name,
        config: n.config,
      };

      // For ifelse nodes, include conditions
      if (n.type === 'ifelse') {
        nodeData.conditions = (n.config?.conditions as IfElseCondition[]) || [];
      }

      // For approval nodes, include message
      if (n.type === 'approval') {
        nodeData.approval_message = (n.config?.message as string) || '';
      }

      return nodeData;
    });

  return {
    name: workflowName,
    agents,
    topologies,
    connections: connectionData,
    nodes: workflowNodes,
  };
};

// API functions
export const workflowApi = {
  // List all workflows
  list: async (): Promise<WorkflowData[]> => {
    const response = await fetch(`${API_BASE}/workflows`);
    if (!response.ok) throw new Error('Failed to fetch workflows');
    return response.json();
  },

  // Get a specific workflow
  get: async (id: string): Promise<WorkflowData> => {
    const response = await fetch(`${API_BASE}/workflows/${id}`);
    if (!response.ok) throw new Error('Failed to fetch workflow');
    return response.json();
  },

  // Create a new workflow
  create: async (data: WorkflowData): Promise<WorkflowData> => {
    const response = await fetch(`${API_BASE}/workflows`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(data),
    });
    if (!response.ok) throw new Error('Failed to create workflow');
    return response.json();
  },

  // Update a workflow
  update: async (id: string, data: Partial<WorkflowData>): Promise<WorkflowData> => {
    const response = await fetch(`${API_BASE}/workflows/${id}`, {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(data),
    });
    if (!response.ok) throw new Error('Failed to update workflow');
    return response.json();
  },

  // Delete a workflow
  delete: async (id: string): Promise<void> => {
    const response = await fetch(`${API_BASE}/workflows/${id}`, {
      method: 'DELETE',
    });
    if (!response.ok) throw new Error('Failed to delete workflow');
  },
};

export const executeApi = {
  // Execute workflow with streaming via POST + fetch ReadableStream
  stream: (
    workflowId: string,
    input: string,
    onMessage: (message: ExecutionMessage) => void,
    onComplete: (result: { output: string; turns_used: number }) => void,
    onError: (error: Error) => void,
    onApprovalRequired?: (data: ApprovalRequiredData) => void,
    history?: ConversationMessage[],
  ): AbortController => {
    const controller = new AbortController();

    fetch(`${API_BASE}/execute/${workflowId}`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ input, config_overrides: {}, history: history || [] }),
      signal: controller.signal,
    })
      .then(async (response) => {
        if (!response.ok) {
          throw new Error(`Execution failed: ${response.status}`);
        }
        if (!response.body) {
          throw new Error('No response body');
        }

        const reader = response.body.getReader();
        const decoder = new TextDecoder();
        let buffer = '';

        while (true) {
          const { done, value } = await reader.read();
          if (done) break;

          buffer += decoder.decode(value, { stream: true });
          const lines = buffer.split('\n');
          buffer = lines.pop() || '';

          for (const line of lines) {
            if (line.startsWith('data: ')) {
              const jsonStr = line.slice(6);
              if (jsonStr.trim()) {
                const data = JSON.parse(jsonStr);

                if (data.type === 'message') {
                  onMessage(data.data);
                } else if (data.type === 'complete') {
                  onComplete(data.data);
                } else if (data.type === 'approval_required' && onApprovalRequired) {
                  onApprovalRequired(data.data);
                }
              }
            }
          }
        }
      })
      .catch((error) => {
        if (error.name !== 'AbortError') {
          onError(error);
        }
      });

    return controller;
  },

  // Resume execution after approval decision
  resume: (
    workflowId: string,
    executionId: string,
    decision: 'approve' | 'reject',
    onMessage: (message: ExecutionMessage) => void,
    onComplete: (result: { output: string; turns_used: number }) => void,
    onError: (error: Error) => void,
    onApprovalRequired?: (data: ApprovalRequiredData) => void,
  ): AbortController => {
    const controller = new AbortController();

    fetch(`${API_BASE}/execute/${workflowId}/resume`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ execution_id: executionId, decision }),
      signal: controller.signal,
    })
      .then(async (response) => {
        if (!response.ok) {
          throw new Error(`Resume failed: ${response.status}`);
        }
        if (!response.body) {
          throw new Error('No response body');
        }

        const reader = response.body.getReader();
        const decoder = new TextDecoder();
        let buffer = '';

        while (true) {
          const { done, value } = await reader.read();
          if (done) break;

          buffer += decoder.decode(value, { stream: true });
          const lines = buffer.split('\n');
          buffer = lines.pop() || '';

          for (const line of lines) {
            if (line.startsWith('data: ')) {
              const jsonStr = line.slice(6);
              if (jsonStr.trim()) {
                const data = JSON.parse(jsonStr);

                if (data.type === 'message') {
                  onMessage(data.data);
                } else if (data.type === 'complete') {
                  onComplete(data.data);
                } else if (data.type === 'approval_required' && onApprovalRequired) {
                  onApprovalRequired(data.data);
                }
              }
            }
          }
        }
      })
      .catch((error) => {
        if (error.name !== 'AbortError') {
          onError(error);
        }
      });

    return controller;
  },

  // Execute workflow synchronously
  sync: async (
    workflowId: string,
    input: string,
    configOverrides?: Record<string, unknown>,
  ): Promise<{
    output: string;
    turns_used: number;
    duration_seconds: number;
    messages: ExecutionMessage[];
  }> => {
    const response = await fetch(`${API_BASE}/execute/${workflowId}/sync`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ input, config_overrides: configOverrides || {} }),
    });
    if (!response.ok) throw new Error('Failed to execute workflow');
    return response.json();
  },

  // Preview workflow (no LLM calls)
  preview: async (
    workflowId: string,
    input: string,
  ): Promise<{
    workflow_id: string;
    workflow_name: string;
    agent_count: number;
    topology_count: number;
    topologies: Array<{
      id: string;
      type: string;
      name: string;
      agents: string[];
      edge_count: number;
    }>;
  }> => {
    const response = await fetch(`${API_BASE}/execute/preview/${workflowId}`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ input, config_overrides: {} }),
    });
    if (!response.ok) throw new Error('Failed to preview workflow');
    return response.json();
  },
};

// Log saving API
interface SaveLogData {
  chatId: string;
  workflowName: string;
  messages: Array<{
    id: string;
    role: string;
    content: string;
  }>;
  logEntries: Array<{
    id: string;
    agent: string;
    content: string;
    timestamp: string;
    topologyInfo?: {
      topologyId: string;
      topologyName: string;
      topologyColor: string;
    } | null;
    agentRole?: string | null;
  }>;
  workflowId: string | null;
  createdAt: string;
}

export const logsApi = {
  // Save chat log to file
  save: async (data: SaveLogData): Promise<{ success: boolean; filepath: string }> => {
    console.log('[logsApi] Saving chat log:', data.chatId);
    const response = await fetch(`${API_BASE}/logs/save`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(data),
    });
    if (!response.ok) {
      const error = new Error(`Failed to save log: ${response.status}`);
      console.error('[logsApi] Save failed:', error);
      throw error;
    }
    const result = await response.json();
    console.log('[logsApi] Log saved successfully:', result.filepath);
    return result;
  },
};

File: backend/app/models/workflow.py
from pydantic import BaseModel, Field
from typing import List, Optional, Dict, Any
from datetime import datetime
from enum import Enum

from .agent import AgentConfig
from .topology import TopologyConfig


class WorkflowNodeType(str, Enum):
    """Types of nodes in the workflow graph."""
    START = "start"
    END = "end"
    AGENT = "agent"
    TOPOLOGY = "topology"
    IFELSE = "ifelse"
    APPROVAL = "approval"
    NOTE = "note"


class IfElseConditionModel(BaseModel):
    """A condition branch in an If/Else node."""
    id: str = Field(..., description="Unique identifier for the condition")
    name: Optional[str] = Field(default=None, description="Display name for the condition")
    condition: str = Field(default="", description="Expression to evaluate")


class WorkflowNode(BaseModel):
    """A node in the workflow graph (for graph-based execution)."""
    id: str = Field(..., description="Unique identifier for the node")
    type: WorkflowNodeType = Field(..., description="Type of node")
    name: str = Field(..., description="Display name")
    config: Dict[str, Any] = Field(default_factory=dict, description="Node configuration")
    # For ifelse nodes
    conditions: List[IfElseConditionModel] = Field(default_factory=list, description="Conditions for if/else branching")
    # For approval nodes
    approval_message: Optional[str] = Field(default=None, description="Message to display for approval")


class ApprovalState(BaseModel):
    """State for a pending approval."""
    execution_id: str = Field(..., description="Unique execution ID")
    workflow_id: str = Field(..., description="Workflow ID")
    node_id: str = Field(..., description="Approval node ID")
    message: str = Field(..., description="Approval message")
    context: Dict[str, Any] = Field(default_factory=dict, description="Execution context")
    created_at: datetime = Field(default_factory=datetime.utcnow)


class ResumeRequest(BaseModel):
    """Request to resume execution after approval."""
    execution_id: str = Field(..., description="Execution ID to resume")
    decision: str = Field(..., description="Decision: 'approve' or 'reject'")


class WorkflowConnection(BaseModel):
    """A workflow-level connection between nodes or topologies."""

    id: str = Field(..., description="Unique identifier for the connection")
    from_id: str = Field(..., description="Source node or topology ID")
    to_id: str = Field(..., description="Target node or topology ID")
    from_type: str = Field(default="node", description="Type of source (node or template)")
    from_port: Optional[int] = Field(default=None, description="Port index for branching (ifelse: 0,1,2..., -1 for else; approval: 0=approve, 1=reject)")


class WorkflowDefinition(BaseModel):
    """Complete workflow definition."""

    id: str = Field(..., description="Unique identifier for the workflow")
    name: str = Field(default="Untitled Workflow", description="Name of the workflow")
    agents: List[AgentConfig] = Field(default_factory=list, description="All agents in the workflow")
    topologies: List[TopologyConfig] = Field(default_factory=list, description="All topology templates")
    connections: List[WorkflowConnection] = Field(default_factory=list, description="Workflow-level connections")
    nodes: List[WorkflowNode] = Field(default_factory=list, description="All workflow nodes (start, end, ifelse, approval)")
    created_at: Optional[datetime] = Field(default=None, description="Creation timestamp")
    updated_at: Optional[datetime] = Field(default=None, description="Last update timestamp")

    class Config:
        json_schema_extra = {
            "example": {
                "id": "workflow-1",
                "name": "Research Workflow",
                "agents": [],
                "topologies": [],
                "connections": [],
            }
        }


class WorkflowCreate(BaseModel):
    """Request model for creating a workflow."""

    name: str = Field(default="Untitled Workflow", description="Name of the workflow")
    agents: List[AgentConfig] = Field(default_factory=list)
    topologies: List[TopologyConfig] = Field(default_factory=list)
    connections: List[WorkflowConnection] = Field(default_factory=list)
    nodes: List[WorkflowNode] = Field(default_factory=list)


class WorkflowUpdate(BaseModel):
    """Request model for updating a workflow."""

    name: Optional[str] = None
    agents: Optional[List[AgentConfig]] = None
    topologies: Optional[List[TopologyConfig]] = None
    connections: Optional[List[WorkflowConnection]] = None
    nodes: Optional[List[WorkflowNode]] = None


class ConversationMessage(BaseModel):
    """A message in conversation history."""
    role: str = Field(..., description="Message role: 'user' or 'assistant'")
    content: str = Field(..., description="Message content")


class ExecutionRequest(BaseModel):
    """Request model for executing a workflow."""

    input: str = Field(..., description="User input message")
    history: List[ConversationMessage] = Field(
        default_factory=list,
        description="Previous conversation history for context"
    )
    config_overrides: Dict[str, Any] = Field(
        default_factory=dict,
        description="Optional config overrides (max_turns, timeout, etc.)"
    )


class ExecutionMessage(BaseModel):
    """A message exchanged during execution."""

    id: str = Field(..., description="Unique message ID")
    from_agent: str = Field(..., description="Sending agent ID")
    to_agent: str = Field(..., description="Receiving agent ID")
    content: str = Field(..., description="Message content")
    timestamp: datetime = Field(default_factory=datetime.utcnow)
    metadata: Dict[str, Any] = Field(default_factory=dict)

    class Config:
        json_schema_extra = {
            "example": {
                "id": "msg-1",
                "from_agent": "agent-1",
                "to_agent": "agent-2",
                "content": "Here is my analysis...",
                "timestamp": "2024-01-01T12:00:00Z",
                "metadata": {},
            }
        }


class ExecutionResult(BaseModel):
    """Result of workflow execution."""

    output: str = Field(..., description="Final output")
    turns_used: int = Field(..., description="Number of turns used")
    duration_seconds: float = Field(..., description="Execution duration")
    messages: List[ExecutionMessage] = Field(default_factory=list, description="All messages exchanged")
